public/app/features/scenes/components/layout/SceneGridLayout.tsx

```diff
@@ -0,0 +1,438 @@
+import { css, cx } from '@emotion/css';
+import React, { useMemo } from 'react';
+import ReactGridLayout from 'react-grid-layout';
+import AutoSizer from 'react-virtualized-auto-sizer';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { Icon, useStyles2, useTheme2 } from '@grafana/ui';
+import { DEFAULT_ROW_HEIGHT, GRID_CELL_HEIGHT, GRID_CELL_VMARGIN, GRID_COLUMN_COUNT } from 'app/core/constants';
+
+import { SceneObjectBase } from '../../core/SceneObjectBase';
+import { SceneComponentProps, SceneLayoutState } from '../../core/types';
+
+interface SceneGridLayoutState extends SceneLayoutState {
+  children: Array<SceneGridCell | SceneGridRow>;
+}
+
+type GridCellLayout = {
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+};
+
+export class SceneGridLayout extends SceneObjectBase<SceneGridLayoutState> {
+  static Component = SceneGridLayoutRenderer;
+
+  updateLayout() {
+    this.setState({
+      children: [...this.state.children],
+    });
+  }
+
+  onResizeStop: ReactGridLayout.ItemCallback = (_, o, n) => {
+    const child = this.state.children.find((c) => c.state.key === n.i);
+    if (!child) {
+      return;
+    }
+    child.setState({
+      size: {
+        ...child.state.size,
+        width: n.w,
+        height: n.h,
+      },
+    });
+  };
+
+  // When layout changes, figure out if it's a nested grid layout and update enclosing cell's size if needed
+  onLayoutChange = (layout: ReactGridLayout.Layout[]) => {
+    let enclosingCell;
+    let parent = this.parent;
+    while (parent) {
+      if (parent instanceof SceneGridCell || parent instanceof SceneGridRow) {
+        enclosingCell = parent;
+        break;
+      } else {
+        parent = parent.parent;
+      }
+    }
+
+    // When not nested, we don't need to update anything
+    if (!enclosingCell) {
+      return;
+    }
+
+    // Collect all cells accumulated heights (cell's y position + height)
+    const accH = [];
+    for (let i = 0; i < layout.length; i++) {
+      const c = layout[i];
+      accH.push(c.y + c.h);
+    }
```

Comment:
not sure this can be correct, if you have many small panels side by side won't this add all their heights together?

Reply:
This is basically calculating the max Y of the cells within a grid - this is relevant for nested grids only.

Reply:
ah, was missing the Math.max call later on, the variable name accH made me think it was the accumulated height